src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -192,458 +96,48 @@ void IUtilities::updateImageViaStagingBuffer(
-                        if(filledAnySlicesOrRowsOrBlocksInLayer)
-                            ret = true;
-                    }
-                            
-                    updateCurrentOffsets();
-
-                    return ret;
-                };
-
-                    // There is remaining layers in region that needs copying
-                auto remainingLayersInRegion = region.imageSubresource.layerCount - currentLayerInRegion;
-                if(currentBlockInRow == 0 && currentRowInSlice == 0 && currentSliceInLayer == 0 && remainingLayersInRegion > 0)
-                {
-                    bool success = tryFillRegion();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
-                }
-                // There is remaining slices in layer that needs copying
-                else if (currentBlockInRow == 0 && currentRowInSlice == 0 && currentSliceInLayer > 0)
-                {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillLayer();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
-                }
-                // There is remaining rows in slice that needs copying
-                else if (currentBlockInRow == 0 && currentRowInSlice > 0)
-                {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillSlice();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
-                }
-                // There is remaining blocks in row that needs copying
-                else if (currentBlockInRow > 0)
+                asset::IImage::SBufferCopy nextRegionToCopy = {};
+                if(availableUploadBufferMemory > 0u && regionIterator.advanceAndCopyToStagingBuffer(nextRegionToCopy, availableUploadBufferMemory, currentUploadBufferOffset, m_defaultUploadBuffer->getBufferPointer()))
                 {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillRow();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
+                    anyTransferRecorded = true;
+                    regionsToCopy.push_back(nextRegionToCopy);
                 }
+                else
+                    break;
             }
 
             if(!regionsToCopy.empty())
             {
                 cmdbuf->copyBufferToImage(m_defaultUploadBuffer.get()->getBuffer(), dstImage, dstImageLayout, regionsToCopy.size(), regionsToCopy.data());
```

Comment:
`dstImageLayout` is a nasty argument name, should be `currentImageLayout`

Reply:
sure, vulkan named it this way and I did the same.